packages/opal-common/opal_common/authentication/oauth2.py

```diff
@@ -0,0 +1,161 @@
+        else:
+            self._exact_match_claims = dict(map(lambda x: x.split("="), cfg.split(",")))
+
+        cfg = opal_common_config.OAUTH2_REQUIRED_CLAIMS
+        if cfg is None:
+            self._required_claims = []
+        else:
+            self._required_claims = cfg.split(",")
+
+    @property
+    def enabled(self):
+        return True
+
+    async def token(self):
+        auth = BasicAuth(self._client_id, self._client_secret)
+        data = {"grant_type": "client_credentials"}
+
+        async with AsyncClient() as client:
+            response = await client.post(self._token_url, auth=auth, data=data)
+            return (response.json())["access_token"]
+
+    def __call__(self, authorization: Optional[str] = Header(None)) -> {}:
+        token = get_token_from_header(authorization)
+        return self.verify(token)
+
+    def verify(self, token: str) -> {}:
+        if self._introspect_url is not None:
+            claims = self._verify_opaque(token)
+        else:
+            claims = self._verify_jwt(token)
+
+        self._verify_exact_match_claims(claims)
+        self._verify_required_claims(claims)
+
+        return claims
+
+    def _verify_opaque(self, token: str) -> {}:
+        response = httpx.post(self._introspect_url, data={"token": token})
+
+        if response.status_code != httpx.codes.OK:
+            raise Unauthorized(
+                description=f"invalid status code {response.status_code}"
+            )
+
+        claims = response.json()
+        active = claims.get("active", False)
+        if not active:
+            raise Unauthorized(description="inactive token")
+
+        return claims or {}
+
+    def _verify_jwt(self, token: str) -> {}:
+        public_key = self._jwk_manager.public_key(token)
+
+        verifier = JWTVerifier(
+            public_key=public_key,
+            algorithm=self._jwt_algorithm,
+            audience=self._jwt_audience,
+            issuer=self._jwt_issuer,
+        )
+        claims = verifier.verify(token)
+
+        return claims or {}
+
+    def _verify_exact_match_claims(self, claims):
+        for key, value in self._exact_match_claims.items():
+            if key not in claims:
+                raise Unauthorized(description=f"missing required '{key}' claim")
+            elif claims[key] != value:
+                raise Unauthorized(description=f"invalid '{key}' claim value")
+
+    def _verify_required_claims(self, claims):
+        for claim in self._required_claims:
+            if claim not in claims:
+                raise Unauthorized(description=f"missing required '{claim}' claim")
+
+
+class CachedOAuth2Authenticator(_OAuth2Authenticator):
+    lock = asyncio.Lock()
+
+    def __init__(self, delegate: OAuth2ClientCredentialsAuthenticator) -> None:
+        self._token = None
+        self._exp = None
+        self._exp_margin = opal_common_config.OAUTH2_EXP_MARGIN
+        self._delegate = delegate
+
+    @property
+    def enabled(self):
```

Comment:
No need to have it again if it's the same as in the `_OAuth2Authenticator` super class

Reply:
Fixed.